### PR TITLE
New API for custom labels and annotations.

### LINKED ISSUE #322: Support annotations and labels containing = and , characters
https://github.com/apache-spark-on-k8s/spark/pull/321#discussion_r119480846

### PR DESCRIPTION
**Pull request by mccheah**
Closes #322.

This APi allows for these labels and annotations to have = and , characters, which is hard to accomplish in the old scheme.

### CODE REVIEW

File: docs/running-on-kubernetes.md [hunk 1]
```diff
@@ -525,9 +525,50 @@ from the other deployment modes. See the [configuration page](configuration.html
...
+    value. For example, setting <code>spark.kubernetes.executor.annotation.identifier</code> to <code>myIdentifier</code>
+    will result in the executor pods having an annotation with key <code>identifier</code> and value
+    <code>myIdentifier</code>. Multiple annotations can be added by setting multiple configurations with this prefix.
+  </td>
+</tr>
+<tr>
   <td><code>spark.kubernetes.driver.labels</code></td>
   <td>(none)</td>
   <td>
+    <i>Deprecated.</i> Use <code>spark.kubernetes.driver.label.<labelKey></code> instead.
```
Comment (ash211) [inline docs/running-on-kubernetes.md]: mention here that this is to support label values with `=` and `,` characters

File: docs/running-on-kubernetes.md [hunk 2]
```diff
@@ -537,6 +578,7 @@ from the other deployment modes. See the [configuration page](configuration.html
   <td><code>spark.kubernetes.driver.annotations</code></td>
   <td>(none)</td>
   <td>
+    <i>Deprecated.</i> Use <code>spark.kubernetes.driver.annotation.<annotationKey></code> instead.
```
Comment (ash211) [inline docs/running-on-kubernetes.md]: mention here that this is to support annotation values with `=` and `,` characters

File: docs/running-on-kubernetes.md [hunk 3]
```diff
@@ -545,6 +587,7 @@ from the other deployment modes. See the [configuration page](configuration.html
   <td><code>spark.kubernetes.executor.labels</code></td>
   <td>(none)</td>
   <td>
+    <i>Deprecated.</i> Use <code>spark.kubernetes.executor.label.<labelKey></code> instead.
```
Comment (ash211) [inline docs/running-on-kubernetes.md]: same

File: docs/running-on-kubernetes.md [hunk 4]
```diff
@@ -554,6 +597,7 @@ from the other deployment modes. See the [configuration page](configuration.html
   <td><code>spark.kubernetes.executor.annotations</code></td>
   <td>(none)</td>
   <td>
+    <i>Deprecated.</i> Use <code>spark.kubernetes.executor.annotation.<annotationKey></code> instead.
```
Comment (ash211) [inline docs/running-on-kubernetes.md]: same

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 5]
```diff
@@ -82,17 +82,36 @@ private[spark] class Client(
   def run(): Unit = {
     validateNoDuplicateFileNames(sparkJars)
     validateNoDuplicateFileNames(sparkFiles)
-    val parsedCustomLabels = ConfigurationUtils.parseKeyValuePairs(
+
+    val parsedCustomLabelsDeprecated = ConfigurationUtils.parseKeyValuePairs(
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: put `driver` in the variable name, e.g. `parsedDriverCustomLabelsDeprecated`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 6]
```diff
@@ -82,17 +82,36 @@ private[spark] class Client(
...
-      s" operations.")
-    val parsedCustomAnnotations = ConfigurationUtils.parseKeyValuePairs(
-      customAnnotations, KUBERNETES_DRIVER_ANNOTATIONS.key, "annotations")
-    require(!parsedCustomAnnotations.contains(SPARK_APP_NAME_ANNOTATION), s"Annotation with key" +
-      s" $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for Spark bookkeeping" +
-      s" operations.")
-    val allLabels = parsedCustomLabels ++ Map(
+    // Remark: getAllWithPrefix strips out the prefix in the returned array.
+    val customLabelsFromPrefixConf = sparkConf.getAllWithPrefix(KUBERNETES_DRIVER_LABEL_PREFIX)
+    val allCustomLabels = parsedCustomLabelsDeprecated.toSeq ++ customLabelsFromPrefixConf
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: `allCustomDriverLabels`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 7]
```diff
@@ -82,17 +82,36 @@ private[spark] class Client(
...
-      s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
-      s" operations.")
-    val parsedCustomAnnotations = ConfigurationUtils.parseKeyValuePairs(
-      customAnnotations, KUBERNETES_DRIVER_ANNOTATIONS.key, "annotations")
-    require(!parsedCustomAnnotations.contains(SPARK_APP_NAME_ANNOTATION), s"Annotation with key" +
-      s" $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for Spark bookkeeping" +
-      s" operations.")
-    val allLabels = parsedCustomLabels ++ Map(
+    // Remark: getAllWithPrefix strips out the prefix in the returned array.
+    val customLabelsFromPrefixConf = sparkConf.getAllWithPrefix(KUBERNETES_DRIVER_LABEL_PREFIX)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: `customDriverLabelsFromPrefixConf`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 8]
```diff
@@ -82,17 +82,36 @@ private[spark] class Client(
...
+    allCustomLabels.groupBy(_._1).foreach {
+      case (key, values) =>
+        require(values.size == 1,
+            s"Cannot have multiple values for a label key, got key $key with values $values")
+    }
+    require(!allCustomLabels.map(_._1).contains(SPARK_APP_ID_LABEL), s"Label with key " +
+        s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
+        s" operations.")
+
+    val parsedCustomAnnotationsDeprecated = ConfigurationUtils.parseKeyValuePairs(
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: `parsedCustomDriverAnnotationsDeprecated`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 9]
```diff
@@ -82,17 +82,36 @@ private[spark] class Client(
...
+    allCustomAnnotations.groupBy(_._1).foreach {
+      case (key, values) =>
+        require(values.size == 1,
+            s"Cannot have multiple values for the same annotation key, got key $key" +
+            s"with values $values")
+    }
+    require(!allCustomAnnotations.map(_._1).contains(SPARK_APP_NAME_ANNOTATION),
+        s"Annotation with key $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for" +
+        s" Spark bookkeeping operations.")
+    val allLabels = allCustomLabels.toMap ++ Map(
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: This is getting long -- we should extract out a method that produces both `allLabels` and `allAnnotations` (named as `driverLabels` and `driverAnnotations`)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 10]
```diff
@@ -82,17 +82,36 @@ private[spark] class Client(
...
+    val customLabelsFromPrefixConf = sparkConf.getAllWithPrefix(KUBERNETES_DRIVER_LABEL_PREFIX)
+    val allCustomLabels = parsedCustomLabelsDeprecated.toSeq ++ customLabelsFromPrefixConf
+    allCustomLabels.groupBy(_._1).foreach {
+      case (key, values) =>
+        require(values.size == 1,
+            s"Cannot have multiple values for a label key, got key $key with values $values")
+    }
+    require(!allCustomLabels.map(_._1).contains(SPARK_APP_ID_LABEL), s"Label with key " +
+        s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
+        s" operations.")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: also should log a warning when using the deprecated config saying that it's deprecated and should be replaced with the new form (labels and annotations for drivers and executors)

File: docs/running-on-kubernetes.md (hunk 1 above)
Comment (mccheah) [reply docs/running-on-kubernetes.md]: Done

File: docs/running-on-kubernetes.md (hunk 2 above)
Comment (mccheah) [reply docs/running-on-kubernetes.md]: Done

File: docs/running-on-kubernetes.md (hunk 3 above)
Comment (mccheah) [reply docs/running-on-kubernetes.md]: Done

File: docs/running-on-kubernetes.md (hunk 4 above)
Comment (mccheah) [reply docs/running-on-kubernetes.md]: Done

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/ClientV2Suite.scala [hunk 11]
```diff
@@ -94,8 +99,11 @@ class ClientV2Suite extends SparkFunSuite with BeforeAndAfter {
       .set(DRIVER_DOCKER_IMAGE, CUSTOM_DRIVER_IMAGE)
       .set(org.apache.spark.internal.config.DRIVER_MEMORY, DRIVER_MEMORY_MB.toLong)
       .set(KUBERNETES_DRIVER_MEMORY_OVERHEAD, DRIVER_MEMORY_OVERHEAD_MB.toLong)
-      .set(KUBERNETES_DRIVER_LABELS, s"$CUSTOM_LABEL_KEY=$CUSTOM_LABEL_VALUE")
-      .set(KUBERNETES_DRIVER_ANNOTATIONS, s"$CUSTOM_ANNOTATION_KEY=$CUSTOM_ANNOTATION_VALUE")
+      .set(KUBERNETES_DRIVER_LABELS, s"$DEPRECATED_CUSTOM_LABEL_KEY=$DEPRECATED_CUSTOM_LABEL_VALUE")
+      .set(KUBERNETES_DRIVER_ANNOTATIONS,
+          s"$DEPRECATED_CUSTOM_ANNOTATION_KEY=$DEPRECATED_CUSTOM_ANNOTATION_VALUE")
+      .set(s"$KUBERNETES_DRIVER_LABEL_PREFIX$CUSTOM_LABEL_KEY", CUSTOM_LABEL_VALUE)
+      .set(s"$KUBERNETES_DRIVER_ANNOTATION_PREFIX$CUSTOM_ANNOTATION_KEY", CUSTOM_ANNOTATION_VALUE)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/ClientV2Suite.scala:106]: @mccheah is it reasonable to check the executor labels/annotations here too?

Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/ClientV2Suite.scala:106]: We don't parse those until we hit `KubernetesClusterSchedulerBackend`.

Comment (ash211) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/ClientV2Suite.scala:106]: ok then, this seems good then

**Review by ash211 — APPROVED** [review]
LGTM, let's merge when build passes

**Pull request merged**